File: php/src/clock.php
<?php

declare(strict_types=1);

function ticks(): int
{
    return (int)(microtime(true) * 1000);
}

class Clock
{
    private int $cycle;
    private int $frame;
    private bool $turbo;
    private int $frames;
    private int $profile;
    private Buttons $buttons;
    private int $start;
    private int $last_frame_start;

    public function __construct(Buttons $buttons, int $frames, int $profile, bool $turbo)
    {
        $this->cycle = 0;
        $this->frame = 0;
        $this->last_frame_start = ticks();
        $this->start = ticks();
        $this->buttons = $buttons;
        $this->frames = $frames;
        $this->profile = $profile;
        $this->turbo = $turbo;
    }

    public function tick(): void
    {
        $this->cycle++;

        // Do a whole frame's worth of sleeping at the start of each frame
        if ($this->cycle % 17556 == 20) {
            // Sleep if we have time left over
            $time_spent = (ticks() - $this->last_frame_start);
            $sleep_for = (1000 / 60) - $time_spent;
            if ($sleep_for > 0 && !$this->turbo && !$this->buttons->turbo) {
                time_nanosleep(0, (int)($sleep_for * 1000000));
            }
            $this->last_frame_start = ticks();

            // Exit if we've hit the frame or time limit
            $duration = (float)($this->last_frame_start - $this->start) / 1000.0;
            if (($this->frames != 0 && $this->frame >= $this->frames) || ($this->profile != 0 && $duration >= $this->profile)) {
                throw new Timeout($this->frame, $duration);
            }

            $this->frame++;
        }
    }
}

File: nim/src/clock.nim
import std/os
import std/times
import std/strformat

import buttons
import errors

# FIXME: use getTime() instead of epochTime()
type
    Clock* = ref object
        buttons: buttons.Buttons
        frames: int
        profile: int
        turbo: bool
        cycle: int
        frame: int
        lastFrameStart: float
        start: float

proc create*(buttons: Buttons, frames: int, profile: int, turbo: bool): Clock =
    return Clock(
      buttons: buttons,
      frames: frames,
      profile: profile,
      turbo: turbo,
      cycle: 0,
      frame: 0,
      last_frame_start: epochTime(),
      start: epochTime(),
    )

proc tick*(self: Clock) =
    self.cycle += 1

    if self.cycle mod 17556 == 20:
        # Sleep if we have time left over
        let timeSpent = epochTime() - self.last_frame_start
        let timePerFrame = 1000 / 60
        if not self.turbo and not self.buttons.turbo and time_spent < time_per_frame:
            let sleepTime = timePerFrame - time_spent
            os.sleep(sleep_time.int)
        self.lastFrameStart = epochTime()

        # Exit if we've hit the frame or time limit
        let duration = self.lastFrameStart - self.start
        if (self.frames != 0 and self.frame >= self.frames) or (self.profile != 0 and duration >= (float)self.profile):
            raise errors.Timeout.newException(
                fmt"Emulated {self.frame:5} frames in {duration:5.2f}s ({(self.frame.float/duration).int}fps)"
            )

        self.frame += 1
